src/bufr/BufrParser/Query/Tokenizer.h

```diff
@@ -0,0 +1,386 @@
+    class FilterToken : public TokenBase<FilterToken> {
+     public:
+        constexpr static const char* Pattern = "\\{\\d+(\\-\\d+)?(\\,\\d+(\\-\\d+)?)*\\}";
+        constexpr static const char* DebugStr = "<filter>";
+
+        std::vector<size_t> indices() { return indices_; }
+
+        /// \brief Tokenize the str_ into tokens.
+        void tokenize() final
+        {
+            static const auto filterRegex =
+                std::regex("\\{(\\d+(\\-\\d+)?)(,\\d+(\\-\\d+)?)*\\}");
+
+            std::set<size_t> indices;
+
+            // match index
+            std::smatch matches;
+            if (std::regex_match(str_, matches, filterRegex))
+            {
+                // regular expression pattern to match ranges and standalone numbers
+                static const std::regex pattern("(\\d+)-(\\d+)|(\\d+)");
+
+                std::smatch match;
+
+                // iterate over all matches in the input string
+                for (auto it = std::sregex_iterator(str_.begin(), str_.end(), pattern);
+                     it != std::sregex_iterator();
+                     ++it)
+                {
+                    // check which capture group matched
+                    if ((*it)[1].matched && (*it)[2].matched)  // range match
+                    {
+                        int start = std::stoi((*it)[1].str());
+                        int end = std::stoi((*it)[2].str());
+                        for (int i = start; i <= end; i++)
+                        {
+                            indices.insert(i);
+                        }
+                    }
+                    else  // standalone number match
+                    {
+                        int number = std::stoi((*it)[3].str());
+                        indices.insert(number);
+                    }
+                }
+            }
+            else
+            {
+                throw eckit::BadParameter("FilterToken::indices: invalid filter: " + str_);
+            }
+
+            indices_ = std::vector<size_t>(indices.begin(), indices.end());
+        }
+
+        /// \brief Get the debug string for the token.
+        std::string debugStr() const override
+        {
+            std::ostringstream debugStr;
+
+            debugStr << "<filter=";
+            for (const auto& index : indices_)
+            {
+                debugStr << index;
+                if (index != indices_.back()) debugStr << ",";
+            }
+            debugStr << ">";
+
+            return debugStr.str();
+        }
+
+     private:
+        std::vector<size_t> indices_;
+    };
+
+    /// \brief Token for a query (i.e. '*/BRIT{2-4}/PCCF[2]'). This token contains tokens for all
+    ///        the elements of the query string.
+    class QueryToken : public TokenBase<QueryToken> {
+     public:
+        constexpr static const char* Pattern =
+            "([A-Z0-9_\\*\\/]+(\\[\\d+\\])?+(\\{[0-9\\-\\,]+\\})?)+";
```

Comment:
this one too, I have a regex check that doesn't like the forward slash in from of the right brack

```suggestion
            "([A-Z0-9_\\*\\]+(\\[\\d+\\])?+(\\{[0-9\\-\\,]+\\})?)+";
```

when I change this it does change the error but still failing:

97: libc++abi: terminating with uncaught exception of type std::__1::regex_error: The parser did not consume the entire regular expression.

Reply:
I'd be careful with that regex checker... What tool is it? What variant of regex parser is it checking against? There are unfortunately a lot of regex parsers with different quirks( for example the python regex parser might have different quirks than the PHP regex parser, etc). That forward slash there is pretty damn important :). No way any of this is going to work without that. Also you just escaped the `]` character...

I'll try to sanatize my regex strings a little (make them compatible with more regex parsers). I use tools like https://regex101.com to play around with these. If you try this don't forget to remove the extra `\` characters so for example the string above would become `([A-Z0-9_\*\/]+((\[\d+\])?)+(\{[0-9\-\,]+\})?)+`. You have to escape the escape character in the c++ strings ;).

@BenjaminRuston Are you able to reproduce @srherbener exactly (same error message)?

Reply:
If anyone can reproduce @srherbener's regex problem please try doing a `git pull` and give my latest version a try. I went through and a made my regex strings play better with more types of regex interpreters. I think I might have found one thing that might have caused his error.

Reply:
@BenjaminRuston Thanks Ben! I spent way to much time trying to recreate the issue on my mahcine... ugg. Thanks for your help :)